### PR TITLE
Optimise performance

### PR DESCRIPTION
**Pull request by pausz**
### Brief description
Changes in this PR include:
-  increase in running time performance of baseline simulations (~10-12% faster) 
    - (#243) 
-  increase in running time performance of simulations with empowerment
-  small refactor of education dynamics methods:
   - method called `update_education()` --> renamed to `advance_education()`
   - the [new `update_education`](https://github.com/fpsim/fpsim/blob/optimise-performance/fpsim/sim.py#L702)  gathers all the necessary education methods, so in `people.update()` we now simply have
   ```python
        # Update education
        if self.pars['education'] is not None:
            alive_now_f = self.filter(self.is_female)
            alive_now_f.update_education()
   ```
- a new [a single function under utils.py](https://github.com/fpsim/fpsim/blob/optimise-performance/fpsim/utils.py#L16) that replaces duplicated code to perform age-matching age_low <= age < age_high  
- an update to [benchmark.py](https://github.com/fpsim/fpsim/blob/optimise-performance/tests/benchmark.py), fixing bugs and updating the commented relative-usage percentages 
    -  (#223)
    -  (#224)
- a [new benchmark script](https://github.com/fpsim/fpsim/blob/optimise-performance/tests/benchmark_empowerment.py) to facilitate comparison between sims with and without empowerment attributes.

### Type(s) of change
- [x] Bugfix
- [x] Refactor
- [ ] New feature
- [ ] Other

### Checklist
- My code follows the [style guide](https://github.com/amath-idm/styleguide)
  - [x] Yes
  - [ ] N/A
- I've commented my code
  - [x] Yes
  - [ ] N/A
- I've incremented the version number
  - [ ] Yes
  - [x] N/A
- I've updated the changelog
  - [ ] Yes
  - [ ] N/A
- I've added tests and checked code coverage
  - [x] Yes
  - [ ] N/A

### CODE REVIEW

File: fpsim/base.py [hunk 1]
```diff
@@ -322,17 +321,18 @@ def filter(self, criteria=None, inds=None):
         # Create a new People object with the same properties as the original
         filtered = object.__new__(self.__class__) # Create a new People instance
         BasePeople.__init__(filtered) # Perform essential initialization
-        filtered.__dict__ = {k:v for k,v in self.__dict__.items()} # Copy pointers to the arrays in People
+        filtered.__dict__ = sc.cp(self.__dict__) # Copy pointers to the arrays in People
```
Comment (cliffckerr) [inline fpsim/base.py:324]: I think even simpler would just be `filtered.__dict__ = self.__dict__.copy()`, right? (https://stackoverflow.com/questions/2465921/how-to-copy-a-dictionary-and-only-edit-the-copy) My bad with the original version of this!

**Review by cliffckerr — APPROVED** [review]
Looks great @pausz, just one small comment

File: fpsim/base.py (hunk 1 above)
Comment (pausz) [reply fpsim/base.py:324]: @cliffckerr I also tried that! but it was slower than using sc.cp/copy.copy -- I can't remember the exact number but could re-rerun and give estimates.

Comment (pausz) [reply fpsim/base.py]: Actually it seems that i was wrong, that'll teach me to do with n=3. I think sc.cp may have seemed faster before i introduced other changes :thinking:. In any case,  I ran benchmarks in a more systematic way:

##### Without empowerment:
- With `dict.copy()`: Average runtime (over 20 runs): 6.2778 s
- With `sc.cp()`: Average runtime (over 20 runs): 6.398 s

##### With empowerment:
- With `dict.copy()`: Average runtime (over 20 runs): 6.7289 s
- With `sc.cp()`: Average runtime (over 20 runs): 6.88505 s

**Pull request merged**